### PR TITLE
Border.add, improve BorderSide.merge

### PR DESCRIPTION
**Pull request by Hixie**
...and other minor Border improvements.
And tests.

This changes the merge logic I added yesterday to not support nulls
but instead support BorderSide.none and equivalents. This makes more
sense when dealing with actual Borders.

### CODE REVIEW

File: packages/flutter/lib/src/painting/borders.dart [hunk 1]
```diff
@@ -374,13 +410,23 @@ class Border {
     return true;
   }
 
+  Border add(Border typedOther) {
```
Comment (tvolkert) [inline packages/flutter/lib/src/painting/borders.dart]: Docs please

File: packages/flutter/lib/src/painting/borders.dart [hunk 2]
```diff
@@ -374,13 +410,23 @@ class Border {
...
+    if (BorderSide.canMerge(top, typedOther.top) &&
+        BorderSide.canMerge(right, typedOther.right) &&
+        BorderSide.canMerge(bottom, typedOther.bottom) &&
+        BorderSide.canMerge(left, typedOther.left)) {
+      return Border.merge(this, typedOther);
+    }
+    return null;
```
Comment (tvolkert) [inline packages/flutter/lib/src/painting/borders.dart]: Why not assert versus returning null?

File: packages/flutter/lib/src/painting/borders.dart [hunk 3]
```diff
@@ -422,17 +468,22 @@ class Border {
     BorderRadius borderRadius,
   }) {
     if (isUniform) {
-      if (shape == BoxShape.circle) {
-        assert(borderRadius == null, 'A borderRadius can only be given for rectangular boxes.');
-        _paintUniformBorderWithCircle(canvas, rect);
-        return;
+      switch (top.style) {
+        case BorderStyle.none:
```
Comment (tvolkert) [inline packages/flutter/lib/src/painting/borders.dart]: Can we have `top.style` be `none` but have a solid style on another edge?

File: packages/flutter/lib/src/painting/borders.dart (hunk 1 above)
Comment (Hixie) [reply packages/flutter/lib/src/painting/borders.dart]: Docs are here and will be merged in a subsequent PR, I'm just merging that branch piecemeal:
https://github.com/Hixie/flutter/blob/borders/packages/flutter/lib/src/painting/borders.dart#L279

File: packages/flutter/lib/src/painting/borders.dart (hunk 2 above)
Comment (Hixie) [reply packages/flutter/lib/src/painting/borders.dart]: In the full branch, if add() returns null then a CompoundBorder is generated instead:
https://github.com/Hixie/flutter/blob/borders/packages/flutter/lib/src/painting/borders.dart#L300

File: packages/flutter/lib/src/painting/borders.dart (hunk 3 above)
Comment (Hixie) [reply packages/flutter/lib/src/painting/borders.dart]: not if isUniform is true

**Review by tvolkert — APPROVED** [review]
LGTM once Travis is happy.

**Pull request merged**